client/client-multi/client-c/src/kaa/platform-impl/common/common_key_utils.c

```diff
@@ -0,0 +1,391 @@
+                             uint8_t *output, size_t *output_payload_size)
+{
+    if (!input || !input_size || !output || !output_payload_size) {
+      return KAA_ERR_BADPARAM;
+    }
+
+    size_t tmp_in_size = input_size;
+    while (input_size >= KAA_SESSION_KEY_LENGTH) {
+        if (aes_encrypt_decrypt(MBEDTLS_AES_DECRYPT, input,
+                                KAA_SESSION_KEY_LENGTH, output,
+                                keys.session_key) < 0) {
+            return KAA_ERR_BADDATA;
+        }
+        input_size -= KAA_SESSION_KEY_LENGTH;
+
+        input += KAA_SESSION_KEY_LENGTH;
+        output += KAA_SESSION_KEY_LENGTH;
+    }
+
+    /* Reduce PKCS7 padding */
+    uint8_t padding_length = *(output - 1);
+    *output_payload_size = tmp_in_size - padding_length;
+
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t ext_get_signature(uint8_t *input, size_t input_size,
+                              uint8_t **output, size_t *output_size)
+{
+    if (!input || !input_size || !output || !output_size) {
+      return KAA_ERR_BADPARAM;
+    }
+
+    kaa_error_t error = rsa_sign(&pk_context_, input, input_size, keys.signature, output_size);
+
+    if (error) {
+      return KAA_ERR_BADDATA;
+    }
+    *output = keys.signature;
+    *output_size = KAA_SIGNATURE_LENGTH;
+
+    return KAA_ERR_NONE;
+}
+
+/*
+ * Generates AES Key.
+ *
+ * bytes The size of the key.
+ * key   Generated AES Key.
+ *
+ */
+static int init_aes_key(unsigned char *key, size_t bytes)
+{
+    mbedtls_ctr_drbg_context ctr_drbg;
+    mbedtls_entropy_context entropy;
+
+    char *pers = "aes_generate_key";
+    int ret;
+
+    mbedtls_entropy_init(&entropy);
+    mbedtls_ctr_drbg_init(&ctr_drbg);
+
+    if ((ret = mbedtls_ctr_drbg_seed(&ctr_drbg, mbedtls_entropy_func, &entropy,
+                                     (unsigned char *) pers, strlen(pers))) != 0) {
+      printf(" failed\n ! mbedtls_ctr_drbg_init returned -0x%04x\n", -ret);
+      goto exit;
+    }
+
+    if ((ret = mbedtls_ctr_drbg_random(&ctr_drbg, key, bytes)) != 0) {
+      printf(" failed\n ! mbedtls_ctr_drbg_random returned -0x%04x\n", -ret);
+      goto exit;
+    }
+
+exit:
+    mbedtls_ctr_drbg_free(&ctr_drbg);
+    mbedtls_entropy_free(&entropy);
+
+    return ret;
+}
+
+static int aes_encrypt_decrypt(int mode, const unsigned char *input, int input_size,
+                               unsigned char *output, const unsigned char *key)
+{
+    if ((input_size % 16) != 0) {
+      return -1;
+    }
+
+    if (mode != MBEDTLS_AES_ENCRYPT && mode != MBEDTLS_AES_DECRYPT) {
+        return -1;
+    }
+
+    static bool initialized = false;
+    static mbedtls_aes_context aes_ctx;
+
+    if (!initialized) {
+        mbedtls_aes_init(&aes_ctx);
+        initialized = true;
+    }
+
+    if (mode == MBEDTLS_AES_ENCRYPT) {
+        mbedtls_aes_setkey_enc(&aes_ctx, key, KAA_SESSION_KEY_LENGTH * 8);
+        mbedtls_aes_crypt_ecb(&aes_ctx, MBEDTLS_AES_ENCRYPT, input, output);
+        return 0;
```

Comment:
Either remove `return 0` and do `else-if` block or remove consequent `f (mode == MBEDTLS_AES_DECRYPT) {` since it is redundant